src/Simplex/Messaging/Agent/Transmission.hs

```diff
@@ -103,60 +107,82 @@ data SMPMessage
         previousMsgHash :: ByteString,
         agentMessage :: AMessage
       }
+  deriving (Show)
 
 data AMessage where
   HELLO :: VerificationKey -> AckMode -> AMessage
   REPLY :: SMPQueueInfo -> AMessage
   A_MSG :: MsgBody -> AMessage
+  deriving (Show)
 
 parseSMPMessage :: ByteString -> Either ErrorType SMPMessage
-parseSMPMessage _ = Left INTERNAL
+parseSMPMessage =
+  first (const $ SYNTAX errBadMessage)
+    . A.parseOnly (smpMessageP <* A.endOfLine <* A.endOfInput)
+  where
+    smpMessageP :: Parser SMPMessage
+    smpMessageP =
+      smpConfirmationP <* A.endOfLine
+        <|> A.endOfLine *> smpClientMessageP
+
+    smpConfirmationP :: Parser SMPMessage
+    smpConfirmationP = SMPConfirmation <$> ("KEY " *> base64P <* A.endOfLine)
+
+    smpClientMessageP :: Parser SMPMessage
+    smpClientMessageP =
+      SMPMessage
+        <$> A.decimal
+        <*> (A.space *> tsIso8601P)
+        <*> (A.space *> base64P)
+        <*> (A.endOfLine *> agentMessageP)
+
+    tsIso8601P :: Parser UTCTime
+    tsIso8601P = maybe (fail "timestamp") pure . parseISO8601 . B.unpack =<< A.takeTill (== ' ')
 
 serializeSMPMessage :: SMPMessage -> ByteString
 serializeSMPMessage = \case
-  SMPConfirmation sKey -> "KEY " <> sKey <> "\r\n\r\n"
+  SMPConfirmation sKey -> smpMessage ("KEY " <> encode sKey) "" ""
   SMPMessage {agentMsgId, agentTimestamp, previousMsgHash, agentMessage} ->
-    "\r\n" <> messageHeader agentMsgId agentTimestamp previousMsgHash <> "\r\n" <> serializeAgentMessage agentMessage
+    let header = messageHeader agentMsgId agentTimestamp previousMsgHash
+        body = serializeAgentMessage agentMessage
+     in smpMessage "" header body
+  where
+    messageHeader msgId ts prevMsgHash =
+      B.unwords [B.pack $ show msgId, B.pack $ formatISO8601Millis ts, encode prevMsgHash]
+    smpMessage smpHeader aHeader aBody = B.intercalate "\n" [smpHeader, aHeader, aBody, ""]
+
+agentMessageP :: Parser AMessage
+agentMessageP = hello <|> reply <|> a_msg
+  where
+    hello = "HELLO " *> (HELLO <$> base64P <*> ackMode)
+    reply = "REPLY " *> (REPLY <$> smpQueueInfoP)
+    a_msg = do
+      size <- "MSG " *> A.decimal :: Parser Int
+      A_MSG <$> (A.endOfLine *> A.take size <* A.endOfLine)
+    ackMode = " NO_ACK" $> AckMode Off <|> pure (AckMode On)
```

Comment:
apply same approach as in `parseCommandP` (#288) to avoid excessive backtracking? or is it unnecessary here?

Reply:
also I guess same could be done for `smpMessageP` (#125). Again not sure if it's needed there, there're only two alternatives rn

Reply:
I am wondering if it has any effect on parsing effectiveness/backtracking - not sure... The main reason for parseCommandP change was code readability (although that applies here as well) - I was actually thinking to just remove simple functions and push it to the main function...

Reply:
when we refactor in the end we need to bring these parts closer - they are too far - difficult to manage